### PR TITLE
Issue 5279: static imports can be separated by local groups

### PR DESCRIPTION
**Pull request by pbludov**
Issue #5279

continuation of #5346

### CODE REVIEW

**Comment by pbludov** [comment]
[checkstyle] [ERROR] C:\projects\checkstyle\.ci\pitest.sh:48: Line should not be longer then 100 symbols [RegexpSingleline]
>please rebase on the latest master

done. Now it fails with
[checkstyle] [ERROR] C:\projects\checkstyle\.ci\pitest.sh:49: Line should not be longer then 100 symbols [RegexpSingleline]
[checkstyle] [ERROR] C:\projects\checkstyle\.ci\pitest.sh:50: Line should not be longer then 100 symbols [RegexpSingleline]

**Comment by pbludov** [comment]
Regression report https://pbludov.github.io/issue-5279/

**Comment by rnveach** [comment]
@pbludov pitest error is valid.

**Comment by romani** [comment]
@pbludov ,

regression reports shows that now we do not report cases where statics have to be separated from ordinary imports - this is not we want. 
Cases:
https://pbludov.github.io/issue-5279/04/RxJava/index.html#A1
...
https://pbludov.github.io/issue-5279/04/infinispan/index.html#A1

According to Check logic and `option=top` property, http://checkstyle.sourceforge.net/property_types.html#importOrder , static imports should be separated from simple imports.

To separate static imports from simple imports might be just another holy war.  We will not consider it now at all. But lets be preserve behavior what we had - demand separation of static imports and simple import.

------------

I will start code review ones CIs become be green.
Looks like 4 lines need to be covered by tests. Some line you did not touch, so some other changes in logic might caused this. Command to reproduce on local is `./.ci/pitest.sh pitest-imports` it is printed in CI log at the most top

@pbludov , do you need help with them ?

**Comment by pbludov** [comment]
> pitest error is valid.

Yes, this PR is transitional and some code is redundant. I'm trying to avoid suppressions but so far without success. See Https://github.com/checkstyle/checkstyle/pull/5760#issuecomment-391928025 and below.

**Comment by pbludov** [comment]
> regression reports shows that now we do not report cases where statics have to be separated from ordinary imports - this is not we want.

For this set of imports:
```
package rx.internal.operators;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import java.util.Arrays;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicInteger;
import static org.junit.Assert.assertEquals;

public class OperatorZipIterableTest {
}
```

The master branch reports:
```
Starting audit...
[ERROR] /home/pb/src/test/TestClass.java:12: 'org.junit.Assert.assertEquals' should be separated from previous imports. [ImportOrder]
[ERROR] /home/pb/src/test/TestClass.java:12: Wrong order for 'org.junit.Assert.assertEquals' import. [ImportOrder]
Audit done.
Checkstyle ends with 2 errors.
```

The patch branch reports
```
Starting audit...
[ERROR] /home/pb/src/test/TestClass.java:12: Wrong order for 'org.junit.Assert.assertEquals' import. [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.
```

A single violation versus multiple violations. This is due to 
> While we have no "groupsStatic" property we should treat all statics as one group, so no separation inside is expected and no grouping from simple import should be applied

For more information, see https://github.com/checkstyle/checkstyle/pull/5760#pullrequestreview-118735293.

So, the code 
```
    private void doVisitToken(FullIdent ident, boolean isStatic,
            boolean previous) {
        final String name = ident.getText();
        final int groupIdx = getGroupNumber(name);
        final int line = ident.getLineNo();

        if (groupIdx > lastGroup) {      <<< Note this
            if (!beforeFirstImport && separated && line - lastImportLine < 2
                && !isInSameGroup(groupIdx, isStatic)) {
                log(line, MSG_SEPARATION, name);     <<< and this
            }
        }
```
is unreachable in the transitional state.

For me, the difference between "two violations in the same line" and "one violation" is not important.
The important difference is "some violations" versus "no violations".

Again, if we include the issue https://github.com/checkstyle/checkstyle/issues/5724 to this PR, this can be solved.

**Comment by romani** [comment]
ok, thanks for explanation, looks like can miss some violation in worse case. False positives more critical then false-negatives in this transition updates.

@pbludov , please review regression and confirm that all changes are expected.
I will continue review of regression too. Ones we both confirm we can ask real users to test your update on their closed sources.

File: src/xdocs/config_imports.xml [hunk 1]
```diff
@@ -1423,6 +1423,17 @@ import java.util.stream.IntStream;
             <td>false</td>
             <td>3.2</td>
           </tr>
+          <tr>
+          <td>separatedStaticGroups</td>
+              <td>
+                whether <b>static imports</b> grouped by <b>top</b> or <b>bottom</b> option should
+                be separated by, at least, one blank line or comment and aren't separated
+                internally.
```
Comment (romani) [inline src/xdocs/config_imports.xml]: brrr .... complete confusion.

in issue we have 
> So, we can add a option named separatedStaticImports which means whether static imports grouped by top or bottom option are separated by local groups.

so this option suppose to care only about separation between static groups and do not rule how statics are separated from simple imports. Please read issue description one more time. Am I missing smth ?
Separation between static and simple import is ruled by `separated` option:
```
$ cat Test.java 
import static org.mockito.Matchers.any;
import com.util.Arrays2;
import java.util.Arrays;

public class Test {
}
rivanov@p5510:/var/tmp$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="ImportOrder">
        <property name="option" value="top"/>
        <property name="separated" value="true"/>
      </module>
    </module>
</module>

$ java -jar checkstyle-8.10.1-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:2: 'com.util.Arrays2' should be separated from previous imports. [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="ImportOrder">
        <property name="option" value="top"/>
        <property name="separated" value="false"/>
      </module>
    </module>
</module>

$ java -jar checkstyle-8.10.1-all.jar -c config.xml Test.java 
Starting audit...
Audit done.

```

**Comment by pbludov** [comment]
>"should be separated from previous imports." should not disappear. Looks like top and bottom does not define separation

It does. For the test class
```
import java.util.Set;
import static org.apache.struts2.views.jsp.AbstractUITagTest.normalize;

class TestClass {
}
```
and the config
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ImportOrder">
      <property name="option" value="bottom"/>
      <property name="separatedStaticGroups" value="true"/>
      <property name="separated" value="true"/>
    </module>
  </module>
</module>
```
Note that `name="option" value="bottom"` the output will be
```
[ERROR] TestClass.java:2: 'org.apache.struts2.views.jsp.AbstractUITagTest.normalize' should be separated from previous imports. [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.
```

When static imports are in the right place, they must be separated.
When they are misplaced, they must be moved. The code in the master branch reports two violations:
"Move the import and separate" The second violation is wrong. The import must be separated for the current position. But the current position is incorrect.
At first, the import should be in the right place. Then it should be re-checked for separation.

Premature violation can confuse people. Consider the code
```
import static foo.bar;

import java.util.Set;
import static foo.baz;

class TestClass {
}
```
and the config
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ImportOrder">
      <property name="option" value="top"/>
      <property name="groups" value="java,org"/>
      <property name="separated" value="true"/>
    </module>
  </module>
</module>
```
The code in master branch reports two violations:
```
Starting audit...
[ERROR] TestClass.java:4: 'foo.baz' should be separated from previous imports. [ImportOrder]
[ERROR] TestClass.java:4: Wrong order for 'foo.baz' import. [ImportOrder]
Audit done.
Checkstyle ends with 2 errors.
```
What does it mean? Maybe the code should be
```
import static foo.bar;

import static foo.baz;

import java.util.Set;

class TestClass {
}
```
Here `foo.baz` is separated from the previous imports. And has the right order. Will there be a violation here? You bet!

**Comment by pbludov** [comment]
>Separation between static and simple import is ruled by `separated` option

No. The `option=TOP` or `option=BOTTOM` is in charge. The parameter `separated` is irrelevant for this.

**Comment by romani** [comment]
> The option=TOP or option=BOTTOM is in charge. The parameter separated is irrelevant for this.

Looks like I proved this by example, `separated` value change validation result - https://github.com/checkstyle/checkstyle/pull/5839#discussion_r191638527 is it just another bug/feature ?

Looks like you changing behavior in code of your PR:
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="ImportOrder">
        <property name="option" value="top"/>
        <property name="separated" value="false"/>
      </module>
    </module>
</module>

$ cat Test.java 
import static org.mockito.Matchers.any;
import com.util.Arrays2;
import java.util.Arrays;

public class Test {
}

$ java -jar checkstyle-8.11-SNAPSHOT-Softus-issue-5279-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:2: 'com.util.Arrays2' should be separated from previous imports. [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.
$ java -jar checkstyle-8.10.1-all.jar -c config.xml Test.java 
Starting audit...
Audit done.
```

but as we make new option `separatedStaticGroups` similar to `separated`. if `sepated=true` demands separation of all imports from statics, because even first/last group should be separated at above/below. So it is reasonable that `separatedStaticGroups` should demand separation first/last group should be separated at above/below.
So 
```
      <property name="option" value="top"/>
      <property name="separatedStaticGroups" value="true"/>
      <property name="separated" value="false"/> <!--FALSE -->
```
should demand separation of static-imports from imports.

> "Move the import" and "separate". The second violation is wrong. The import must be separated for the current position. But the current position is incorrect.

ok, you please confirm you skip validation of separation is import in mis-ordered.
It is completely fine, and make sense, just need to be sure that is intentional and in addition to new feature. 

-------------------------

Note for myself:
http://checkstyle.sourceforge.net/property_types.html#importOrder other options like "infow", "above", "under" where groups are the same for imports-static and imports and following option `groups`. So in documentation `separatedStaticGroups` and `groupsStatic` should be clear note that they are active only for "top" "buttom" modes.

**Comment by romani** [comment]
Looks like with this update it is become absolutely create that `top` and `bottom` demands separation. 
If this is true we need to update in this PR http://checkstyle.sourceforge.net/property_types.html#importOrder to clearly define/show expected behavior.

**Comment by pbludov** [comment]
What about this crazy config
```
   <property name="option" value="above"/>
    <property name="separatedStaticGroups" value="true"/>
    <property name="separated" value="true"/>
```
A matching set of imports will be
```
import static foo.bar;
import static foo.baz;
import java.lang.Object;

import org.foo;
```
I'm absolutely sure that no one will use it, but we must support it as well.

Also, when both `separated` and `separatedStaticGroups` are set to `false`, the difference between
`option`=`TOP` and `option`=`ABOVE` is very important.

I'm going to make a review of the `ImportOrderCheck` documentation.
We need an example for each combination of parameters.

**Comment by romani** [comment]
> I'm absolutely sure that no one will use it, but we must support it as well.

we must follow common sense. All other is matter of limitations described in documentation.
In our last changes I tried to lean Checks to support the same as IDEs can support, as 99% of people use IDEs formatting/organizers and use chekstyle to reconcile different IDEs and recheck that IDE was set up properly.

> Also, when both separated and separatedStaticGroups are set to false, the difference between
option=TOP and option=ABOVE is very important.

is it about presence of separator between static-imports and imports ? we could ignore this for now, if somebody complain we can make one more option like `separateStatic=true` make them happy in their very fancy style of imports.

**Comment by pbludov** [comment]
Just to confirm, for `option`=`ABOVE`/`UNDER` and `INFLOW` the upcoming option `staticGroups` should be ignored?

For `option`=`UNDER` and `separatedStaticGroups`=`true` and `separated`=`true` the imports should be
```
        import java.awt.Dialog;

        import static java.awt.Button.A;

        import java.io.File;
        import java.io.IOException;

        import static java.io.File.createTempFile;
```
?

**Comment by romani** [comment]
@pbludov , please add user test cases from #5279 to our test code base.
It is better to add them to `it` tests, as test cases that can not be changed and works on exact config.
As example we have there google style files, changes in them are always an red flag during review, so regression is unlikely to be missed.
Bad updates in UTs are easy to miss in review, it is easy to be mislead that new changes require changes in expectations, this is how we end-up with such regression.  

https://github.com/checkstyle/checkstyle/tree/master/src/it
Test Classes: java/org/checkstyle/test/checks/importorder/....
Test Inputs: resources/com/google/checkstyle/test/checks/importorder/....
It might be good in inputs define from where they come, and on what config they support to work (as multiline comment in Input file), to make it clear that it is NOT an artificial set of imports.

Do you  see this reasonable ?

File: .ci/pitest.sh [hunk 2]
```diff
@@ -31,17 +31,31 @@ function checkPitestReport() {
...
 
-pitest-annotation|pitest-design|pitest-header|pitest-imports \
+pitest-annotation|pitest-design|pitest-header \
 |pitest-metrics|pitest-misc|pitest-modifier|pitest-naming \
 |pitest-regexp|pitest-sizes|pitest-whitespace|pitest-ant \
 |pitest-api|pitest-common|pitest-filters|pitest-main \
 |pitest-packagenamesloader|pitest-tree-walker|pitest-utils \
 |pitest-xpath)
   mvn -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
+  # Till https://github.com/checkstyle/checkstyle/issues/5724
```
[the remark this answers is not in the document]
Comment (pbludov) [reply .ci/pitest.sh]: done

**Comment by romani** [comment]
@pbludov ,

first `it` are already created - https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle , please rebase.

**Comment by pbludov** [comment]
> please add user test cases from #5279 to our test code base.

All of them are possible only with the "staticGroups" property. A'll add them in the scope of the issue https://github.com/checkstyle/checkstyle/issues/5724

**Comment by rnveach** [comment]
@pbludov @romani What is the status of this PR?

**Comment by romani** [comment]
@pbludov , please find me chat to discuss how can we speedup resolving of this issue and how can we help you to complete.

**Comment by pbludov** [comment]
> What is the status of this PR?

The PR is *not* ready to merge. We need more tests for this check.

**Comment by romani** [comment]
@pbludov , it will be better to close this PR and open new PR, with clean history of changes based on latest code in master.

Please do not forget  about my request moving all users input cases to ITs - https://github.com/checkstyle/checkstyle/pull/5839#issuecomment-393399792 to make it clear for future that changes in behavior is not expected as all and if any happen should be considered as serious breaking compatibility.

**Pull request closed without merging**